### PR TITLE
mon: {mon,osd,mds} {versions,count-metadata}

### PR DESCRIPTION
**Pull request by liewegas**
<pre>
gnit:build (wip-versions) 02:46 PM $ bin/ceph osd versions
{
    "ceph version 12.0.2-2073-gafcfc1d (afcfc1d7d99fae1697d687f65ce297de5c94b3a8) luminous (dev)": 3
}
gnit:build (wip-versions) 02:47 PM $ bin/ceph mds count-metadata kernel_version
{
    "4.8.8-300.fc25.x86_64": 1
}
</pre>

also fix an accidental change to the ceph version string (extra space)

### CODE REVIEW

**Comment by liewegas** [comment]
retest this please

**Comment by yuriw** [comment]
this was tested and approved for merge
@liewegas

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -4172,6 +4194,20 @@ bool OSDMonitor::preprocess_command(MonOpRequestRef op)
       f->close_section();
     }
     f->flush(rdata);
+  } else if (prefix == "osd versions") {
+    if (!f)
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: this is not necessary, `f` is always valid. see https://github.com/ceph/ceph/pull/15436/files#diff-0a5db46a44ae9900e226289a810f10e8R3975
```c++
  boost::scoped_ptr<Formatter> f(Formatter::create(format));
```

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -4172,6 +4194,20 @@ bool OSDMonitor::preprocess_command(MonOpRequestRef op)
...
+      f.reset(Formatter::create("json-pretty"));
+    count_metadata("ceph_version", f.get());
+    f->flush(rdata);
+    r = 0;
+  } else if (prefix == "osd count-metadata") {
+    if (!f)
+      f.reset(Formatter::create("json-pretty"));
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: ditto.

File: src/mon/Monitor.cc [hunk 3]
```diff
@@ -3205,6 +3207,24 @@ void Monitor::handle_command(MonOpRequestRef op)
     f->flush(ds);
     rdata.append(ds);
     rs = "";
+  } else if (prefix == "mon versions") {
+    if (!f)
```
Comment (tchaikov) [inline src/mon/Monitor.cc]: this is not necessary either.

```c++
  boost::scoped_ptr<Formatter> f(Formatter::create(format));
```

File: src/mon/MDSMonitor.cc [hunk 4]
```diff
@@ -969,6 +969,20 @@ bool MDSMonitor::preprocess_command(MonOpRequestRef op)
       f->close_section();
     }
     f->flush(ds);
+  } else if (prefix == "mds versions") {
+    if (!f)
```
Comment (tchaikov) [inline src/mon/MDSMonitor.cc]: this is not necessary, either either.

**Comment by tchaikov** [comment]
i will send a cleanup PR as the follow-up of this PR.

File: src/mon/MDSMonitor.cc (hunk 4 above)
Comment (tchaikov) [inline src/mon/MDSMonitor.cc]: sorry, i was wrong, `Formatter::create("plain")` will return a nullptr.

**Pull request merged**